src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/SemanticTokensCache.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+#pragma warning disable CS0618
+#nullable enable
+
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using System.Collections.Generic;
+using System;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
+{
+    /// <summary>
+    /// Cache used by <see cref="DefaultRazorSemanticTokensInfoService"/> to keep track of
+    /// previously computed tokens.
+    /// </summary>
+    internal class SemanticTokensCache<T> where T : SemanticTokens
```

Comment:
Out of curiosity, after moving to a custom data type did you feel it simplified consuming code a lot? As I read through I kept comparing it against a `MemoryCache<Uri, MemoryCache<string, VersionedSemanticTokens>>` construct so wanted to ask your opinion 😄